src/server/pkg/storage/chunk/chunk.proto

```diff
@@ -0,0 +1,14 @@
+syntax = "proto3";
+
+package chunk;
+option go_package = "github.com/pachyderm/pachyderm/src/server/pkg/storage/chunk";
+
+// DataRef is a reference to a chunk of data.
+message DataRef {
+  string hash = 1;
+  // Hash of the subchunk (defined by offset and size).
+  // Should be empty if this is a reference to the whole chunk.
+  string sub_hash = 2;
+  uint64 offset = 3;
+  uint64 size = 4;
+}
```

Comment:
This message seems a bit confusing. Does `size` represent the total size of the chunk being reference or just the sub-chunk? It'd be good to have some more documentation on this and maybe some better names, since this is one of the central messages in the new storage layer. I'm also wondering if maybe this message is doing too much, and should be split into 2 messages, one that just has the hash and size of the chunk, and another that contains the first message and records where the interesting information begins and what the sub hash is.

Reply:
The purpose of this data structure is to be a reference to data, it does not really reflect chunk information outside of the identifier for the containing chunk. I agree that the names could probably use some work though, particularly hash/sub_hash. I am open for suggestions there. It might make sense to have another data structure for chunks, but I am not sure it is necessary if we want chunks to map 1:1 to objects. This is the data structure that is exposed by the chunk storage layer (which does the deduplication), so it is not necessary to know stuff about chunks (besides an identifier) outside of this abstraction.

Reply:
Also, I think it might make more sense not to use `uint64`s here unless it makes interfacing with `uint64`s elsewhere a pain. I've come to the realization that `uint64`s are more trouble than they're worth and just lead to a bunch of casting annoyances. And it's not like the added bit is going to help us with anything, honestly I suspect if we ever do get to sizes that use that bit, a lot of things will break.

Reply:
That seems like a good idea. I will stick to `int64` moving forward.

Reply:
So naming wise, I think it'd make more sense to me if `hash` was renamed to `chunk`, since it's purpose is to tell you which chunk you're referencing. You might even want to make a wrapper message:

```proto
message Chunk {
  string hash = 1;
}
```

Then I think `sub_hash` can be renamed to `hash` which I think makes things simpler and makes it clear that this is the hash that's intended to be used to determine the hashes of files.

Reply:
I think this makes sense. It could be useful to have a Chunk proto in the future anyways.

Reply:
Yeah, despite the fact that it creates a bit more typing I've overall been happy with our decision to use wrapper structs to make the names well-typed.